realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmJsonTypeHelper.java

```diff
@@ -349,52 +343,52 @@ public void emitStreamTypeConversion(String interfaceName, String setter, String
         public void emitGetObjectWithPrimaryKeyValue(
                 String qualifiedRealmObjectClass, String qualifiedRealmObjectProxyClass, String fieldName, JavaWriter writer)
                 throws IOException {
-            throw new IllegalArgumentException("'byte[]' is not allowed as a primary key value.");
+            throw new IllegalArgumentException("'Date' is not allowed as a primary key value.");
         }
     }
 
     private static class MutableRealmIntegerTypeConverter implements JsonToRealmFieldTypeConverter {
         // @formatter:off
-       @Override
-       public void emitTypeConversion(String interfaceName, String getter, String fieldName, String fieldType, JavaWriter writer) throws IOException {
-           writer
-               .beginControlFlow("if (json.has(\"%s\"))", fieldName)
-                   .beginControlFlow("if (json.isNull(\"%s\"))", fieldName)
-                       .emitStatement(Constants.STATEMENT_EXCEPTION_ILLEGAL_NULL_VALUE, fieldName)
-                   .nextControlFlow("else")
-                       .emitStatement("((%s) obj).%s().set(json.getLong(\"%s\"))", interfaceName, getter, fieldName)
-                   .endControlFlow()
-               .endControlFlow();
-       }
-       // @formatter:on
-
-       // @formatter:off
-       @Override
-       public void emitStreamTypeConversion(String interfaceName, String getter, String fieldName, String fieldType, JavaWriter writer, boolean isPrimaryKey) throws IOException {
-           writer
-               .beginControlFlow("if (reader.peek() == JsonToken.NULL)")
-                   .emitStatement("reader.skipValue()")
-                   .emitStatement(Constants.STATEMENT_EXCEPTION_ILLEGAL_NULL_VALUE, fieldName)
-               .nextControlFlow("else")
-                   .emitStatement("((%s) obj).%s().set(reader.nextLong(\"%s\"))", interfaceName, getter, fieldName)
-               .endControlFlow();
-       }
-       // @formatter:on
+        @Override
+        public void emitTypeConversion(String varName, String accessor, String fieldName, String fieldType, JavaWriter writer)
+                throws IOException {
+            writer
+                .beginControlFlow("if (json.has(\"%s\"))", fieldName)
+                    .emitStatement("%1$s.%2$s().set((json.isNull(\"%3$s\")) ? null : json.getLong(\"%3$s\"))", varName, accessor, fieldName)
+                .endControlFlow();
+        }
+        // @formatter:on
+
+        // @formatter:off
+        @Override
+        public void emitStreamTypeConversion(String varName, String accessor, String fieldName, String fieldType, JavaWriter writer, boolean isPrimaryKey)
+                throws IOException {
+            writer
+                .emitStatement("Long val = null")
+                .beginControlFlow("if (reader.peek() != JsonToken.NULL)")
+                    .emitStatement("val = reader.nextLong()")
```

Comment:
In most cases, `val` is assigned twice (first time is `null` and the second time is the value from json).
Assigning twice is not only less efficient, but also less readable.
How about moving assigning `null` to `else` clause?